src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -458,14 +469,18 @@ public static class ScoreOrderedResultRetriever extends AbstractIterator<Unfilte
         private final ReadExecutionController executionController;
         private final QueryContext queryContext;
 
-        private HashSet<PrimaryKey> keysSeen;
-        private HashSet<PrimaryKey> updatedKeys;
+        private final HashSet<PrimaryKey> processedKeys;
+        private final Queue<UnfilteredRowIterator> pendingRows;
+
+        private final int limit;
```

Comment:
This is a soft limit, an optimization hint, right?
The caller of this ResultRetriever can fetch more rows and the ResultRetriever won't end the stream on limit.
Maybe add a word about it in the doc.

Reply:
It actually is a hard limit. I can change it so that we just pull one more row in if we exceed the limit. However, I don't see a scenario where we would or should.

Reply:
Never mind, it was already a soft limit. I'll document the behavior though since that will make it easier to understand the intention.